lib/topological_inventory/providers/common/heartbeat.rb

```diff
@@ -0,0 +1,106 @@
+module TopologicalInventory
+  module Providers
+    module Common
+      class << self
+        attr_writer :heartbeat
+      end
+
+      def self.heartbeat(name)
+        @heartbeat ||= {}
+        @heartbeat[name] ||= TopologicalInventory::Providers::Common::Heartbeat.new(name)
+      end
+
+      module HeartbeatQueue
+        def heartbeat(name)
+          TopologicalInventory::Providers::Common.heartbeat(name)
+        end
+      end
+
+      class Heartbeat
+        HEARTBEAT_TIMEOUT = 10.freeze
+        HEARTBEAT_THREAD_TIMEOUT = 2.freeze
+
+        attr_accessor :heartbeat_queue, :finished, :name
+
+        def initialize(name)
+          self.heartbeat_queue = Concurrent::Array.new([:tick])
+          self.finished = Concurrent::AtomicBoolean.new(false)
+          self.name = name
+
+          self.create_heartbeat_dir
+        end
+
+        def run_thread
+          Thread.new do
+            until finished.value
+              if self.heartbeat_queue.present?
+                self.touch_heartbeat_file
+                self.heartbeat_queue.clear
+              end
+              sleep(HEARTBEAT_THREAD_TIMEOUT)
+            end
+          end
+        end
+
+        def queue_tick
+          self.heartbeat_queue << :tick
+        end
+
+        def stop
+          self.finished = Concurrent::AtomicBoolean.new(true)
+        end
+
+        def touch_heartbeat_file
+          File.write(self.heartbeat_file, (Time.now.utc + HEARTBEAT_TIMEOUT).to_s)
+        end
+
+        def create_heartbeat_dir
+          full_heartbeat_dir = File.expand_path(heartbeat_dir, __FILE__)
+          Dir.mkdir(full_heartbeat_dir) unless File.exist?(full_heartbeat_dir)
+        end
+
+        def heartbeat_dir
+          self.class.heartbeat_dir
+        end
+
+        def heartbeat_file
+          self.class.heartbeat_file(self.name)
+        end
+
+        def self.heartbeat_dir
+          "../../../../tmp"
```

Comment:
I was not able to use Tempfile, because it seems that there now way how to access created tempfile again.

so I used `File.exapand('tmp', '.')` and it looks that it create directory in root of app.